File: GameLibrary/src/cc/lib/zombicide/ZMove.java
package cc.lib.zombicide;

import java.util.Arrays;
import java.util.List;

import cc.lib.game.Utils;
import cc.lib.ui.IButton;
import cc.lib.utils.Reflector;

public class ZMove extends Reflector<ZMove> implements IButton {

    static {
        addAllFields(ZMove.class);
    }

    public final ZMoveType type;
    public final Integer integer;
    public final ZPlayerName character;
    public final ZEquipment equipment;
    public final ZEquipSlot fromSlot;
    public final ZEquipSlot toSlot;
    public final List list;
    public final ZDir dir;
    public final ZSkill skill;

    public ZMove() {
        this(null);
    }

    public ZMove(ZMove copy, Object singleListElement) {
        this(copy.type, copy.integer, copy.character, copy.equipment, copy.fromSlot, copy.toSlot, Arrays.asList(singleListElement), copy.dir, copy.skill);
    }

    public ZMove(ZMove copy, Object singleListElement, int integer) {
        this(copy.type, integer, copy.character, copy.equipment, copy.fromSlot, copy.toSlot, Arrays.asList(singleListElement), copy.dir, copy.skill);
    }

    public ZMove(ZMove copy, Object singleListElement, ZPlayerName character) {
        this(copy.type, copy.integer, character, copy.equipment, copy.fromSlot, copy.toSlot, Arrays.asList(singleListElement), copy.dir, copy.skill);
    }

    public ZMove(ZMove copy, Object singleListElement, ZEquipment equipment) {
        this(copy.type, copy.integer, copy.character, equipment, copy.fromSlot, copy.toSlot, Arrays.asList(singleListElement), copy.dir, copy.skill);
    }

    private ZMove(ZMoveType type) {
        this(type, (Integer)null);
    }

    private ZMove(ZMoveType type, ZDir dir) {
        this(type, null, null, null, null, null, null, dir, null);
    }

    private ZMove(ZMoveType type, Integer num) {
        this(type, num, null, null, null, null, null, null, null);
    }

    private ZMove(ZMoveType type, List list) {
        this(type, null, null, null, null, null, list, null, null);
    }

    private ZMove(ZMoveType type, List list, ZSkill skill) {
        this(type, null, null, null, null, null, list, null, skill);
    }

    private ZMove(ZMoveType type, ZEquipment equip, ZEquipSlot fromSlot) {
        this(type, null, null, equip, fromSlot, null, null, null, null);
    }

    private ZMove(ZMoveType type, int targetIndex, ZPlayerName character, ZEquipment equip, ZEquipSlot fromSlot, ZEquipSlot toSlot, List list) {
        this(type, targetIndex, character, equip, fromSlot, toSlot, list, null, null);
    }

    private ZMove(ZMoveType type, Integer integer, ZPlayerName character, ZEquipment equip, ZEquipSlot fromSlot, ZEquipSlot toSlot, List list, ZDir dir, ZSkill skill) {
        this.type = type;
        this.integer = integer;
        this.character = character;
        this.equipment = equip;
        this.fromSlot = fromSlot;
        this.toSlot = toSlot;
        this.list  = list;
        this.dir = dir;
        this.skill = skill;
    }

    @Override
    public String toString() {
        return "ZMove{" +
                "type=" + type +
                ", integer=" + integer +
                ", equipment=" + equipment +
                ", fromSlot=" + fromSlot +
                ", toSlot=" + toSlot +
                ", list=" + list +
                ", dir=" + dir +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null)
            return false;
        if (!(o instanceof ZMove))
            return false;
        ZMove zMove = (ZMove) o;
        return type == zMove.type
                && dir == zMove.dir
                && Utils.isEquals(equipment, zMove.equipment)
                && Utils.isEquals(integer, zMove.integer)
                && character == zMove.character
                && fromSlot == zMove.fromSlot
                && toSlot == zMove.toSlot
                && skill == zMove.skill;
    }

    @Override
    public String getTooltipText() {
        if (equipment != null) {
            return equipment.getTooltipText();
        }
        if (skill != null) {
            return skill.getTooltipText();
        }
        if (this.character != null) {
            return character.getTooltipText();
        }
        return type.getTooltipText();
    }

    @Override
    public String getLabel() {
        String label = Utils.toPrettyString(type.name());
        if (equipment != null)
            label += " " + equipment.getLabel();
        if (toSlot == null && fromSlot != null) {
            label += " from " + fromSlot.getLabel();
        }
        if (toSlot != null)
            label += " to " + toSlot.getLabel();

        return label;
    }

    static ZMove newEndTurn() {
        return new ZMove(ZMoveType.END_TURN);
    }

    static ZMove newWalkMove(List<Integer> zones) {
        return new ZMove(ZMoveType.WALK, zones);
    }

    static ZMove newJumpMove(List<Integer> zones) {
        return new ZMove(ZMoveType.JUMP, zones);
    }

    static ZMove newChargeMove(List<Integer> zones) {
        return new ZMove(ZMoveType.CHARGE, zones);
    }

    static ZMove newUseLeftHand() {
        return new ZMove(ZMoveType.USE_LEFT_HAND);
    }

    static ZMove newUseRightHand() {
        return new ZMove(ZMoveType.USE_RIGHT_HAND);
    }

    static ZMove newToggleDoor(List<ZDoor> doors) {
        return new ZMove(ZMoveType.OPERATE_DOOR, doors);
    }

    static ZMove newSearchMove(int zoneIndex) {
        return new ZMove(ZMoveType.SEARCH, zoneIndex);
    }

    static ZMove newMeleeAttackMove(List<ZWeapon> weapons) {
        return new ZMove(ZMoveType.MELEE_ATTACK, weapons);
    }

    static ZMove newRangedAttackMove(List<ZWeapon> weapons) {
        return new ZMove(ZMoveType.RANGED_ATTACK, weapons);
    }

    static ZMove newMagicAttackMove(List<ZWeapon> weapons) {
        return new ZMove(ZMoveType.MAGIC_ATTACK, weapons);
    }

    static ZMove newThrowEquipmentMove(List<ZEquipment> slots) {
        return new ZMove(ZMoveType.THROW_ITEM, slots);
    }

    static ZMove newInventoryMove() {
        return new ZMove(ZMoveType.INVENTORY);
    }

    static ZMove newTradeMove(List<ZPlayerName> tradeOptions) {
        return new ZMove(ZMoveType.TRADE, tradeOptions);
    }

    static ZMove newConsumeMove(ZEquipment equip, ZEquipSlot slot) {
        return new ZMove(ZMoveType.CONSUME, equip, slot);
    }

    static ZMove newEquipMove(ZEquipment equip, ZEquipSlot fromSlot, ZEquipSlot toSlot) {
        return new ZMove(ZMoveType.EQUIP, 0, null, equip, fromSlot, toSlot, null);
    }

    static ZMove newKeepMove(ZEquipment equip) {
        return new ZMove(ZMoveType.KEEP, equip, null);
    }

    static ZMove newUnequipMove(ZEquipment equip, ZEquipSlot slot) {
        return new ZMove(ZMoveType.UNEQUIP, equip, slot);
    }

    static ZMove newDisposeMove(ZEquipment equip, ZEquipSlot slot) {
        return new ZMove(ZMoveType.DISPOSE, equip, slot);
    }

    static ZMove newGiveMove(ZPlayerName taker, ZEquipment toGive) {
        return new ZMove(ZMoveType.GIVE, 0, taker, toGive, null, null, null);
    }

    static ZMove newTakeMove(ZPlayerName giver, ZEquipment toTake) {
        return new ZMove(ZMoveType.TAKE, 0, giver, toTake, null, null, null);
    }

    public static ZMove newObjectiveMove(int zone) {
        return new ZMove(ZMoveType.TAKE_OBJECTIVE, zone);
    }

    static ZMove newReloadMove(ZWeapon slot) {
        return new ZMove(ZMoveType.RELOAD, slot, null);
    }

    static ZMove newPickupItemMove(List<ZEquipment> takables) {
        return new ZMove(ZMoveType.PICKUP_ITEM, takables);
    }

    static ZMove newDropItemMove(List<ZEquipment> items) {
        return new ZMove(ZMoveType.DROP_ITEM, items);
    }

    static ZMove newWalkDirMove(ZDir dir) {
        return new ZMove(ZMoveType.WALK_DIR, dir);
    }

    static ZMove newSwitchActiveCharacter() {
        return new ZMove(ZMoveType.SWITCH_ACTIVE_CHARACTER);
    }

    static ZMove newMakeNoiseMove(int occupiedZone) {
        return new ZMove(ZMoveType.MAKE_NOISE, occupiedZone);
    }

    static ZMove newShoveMove(List<Integer> toZones) {
        return new ZMove(ZMoveType.SHOVE, toZones);
    }

    static ZMove newReRollMove() {
        return new ZMove(ZMoveType.REROLL);
    }

    static ZMove newKeepRollMove() {
        return new ZMove(ZMoveType.KEEP_ROLL);
    }

    static ZMove newEnchantMove(List<ZSpell> spells) {
        return new ZMove(ZMoveType.ENCHANT, spells);
    }

    static ZMove newBornLeaderMove(List<ZPlayerName> options) {
        return new ZMove(ZMoveType.BORN_LEADER, options);
    }

    static ZMove newBloodlustMeleeMove(List<Integer> zones, ZSkill skill) {
        return new ZMove(ZMoveType.BLOODLUST_MELEE, zones, skill);
    }

    static ZMove newBloodlustRangedMove(List<Integer> zones, ZSkill skill) {
        return new ZMove(ZMoveType.BLOODLUST_RANGED, zones, skill);
    }

    static ZMove newBloodlustMagicMove(List<Integer> zones, ZSkill skill) {
        return new ZMove(ZMoveType.BLOODLUST_MAGIC, zones, skill);
    }

    static ZMove newDisposeEquipmentMove(ZEquipment e) {
        return new ZMove(ZMoveType.DISPOSE, e, null);
    }

    static ZMove newIgniteMove(List<Integer> ignitableZones) {
        return new ZMove(ZMoveType.IGNITE, ignitableZones);
    }
}

File: Zombicide/src/cc/lib/zombicide/ZMove.kt
package cc.lib.zombicide


import cc.lib.reflector.Reflector
import cc.lib.ui.IButton

data class ZMove(
	val type: ZMoveType = ZMoveType.END_TURN,
	val integer: Int? = null,
	val character: ZPlayerName? = null,
	val equipment: ZEquipment<*>? = null,
	val fromSlot: ZEquipSlot? = null,
	val toSlot: ZEquipSlot? = null,
	val list: List<*>? = null,
	val skill: ZSkill? = null,
	val action: ZActionType? = null,
	val text: String? = null,
	val familiar: ZFamiliarType? = null
) : Reflector<ZMove>(), IButton {
	companion object {
		fun newEndTurn(): ZMove {
			return ZMove(type = ZMoveType.END_TURN)
		}

		fun newWalkMove(zones: List<Int>, action: ZActionType?): ZMove {
			return ZMove(type = ZMoveType.WALK, list = zones, action = action)
		}

		fun newJumpMove(zones: List<Int>): ZMove {
			return ZMove(type = ZMoveType.JUMP, list = zones)
		}

		fun newChargeMove(zones: List<Int>): ZMove {
			return ZMove(type = ZMoveType.CHARGE, list = zones)
		}

		fun newUseSlot(slot: ZEquipSlot): ZMove {
			return ZMove(type = ZMoveType.USE_SLOT, fromSlot = slot)
		}

		fun newToggleDoor(doors: List<ZDoor>): ZMove {
			return ZMove(type = ZMoveType.OPERATE_DOOR, list = doors)
		}

		fun newBarricadeDoor(doors: List<ZDoor>): ZMove {
			return ZMove(type = ZMoveType.BARRICADE, list = doors)
		}

		fun newSearchMove(zoneIndex: Int): ZMove {
			return ZMove(type = ZMoveType.SEARCH, integer = zoneIndex)
		}

		fun newMeleeAttackMove(weapons: List<ZWeapon>): ZMove {
			return ZMove(type = ZMoveType.MELEE_ATTACK, list = weapons, action = ZActionType.MELEE)
		}

		fun newRangedAttackMove(weapons: List<ZWeapon>, zoneIdx: Int? = null): ZMove {
			return ZMove(type = ZMoveType.RANGED_ATTACK, list = weapons, action = ZActionType.RANGED, integer = zoneIdx)
		}

		fun newMagicAttackMove(weapons: List<ZWeapon>, zoneIdx: Int? = null): ZMove {
			return ZMove(type = ZMoveType.MAGIC_ATTACK, list = weapons, action = ZActionType.MAGIC, integer = zoneIdx)
		}

		fun newThrowEquipmentMove(slots: List<ZEquipment<*>?>): ZMove {
			return ZMove(type = ZMoveType.THROW_ITEM, list = slots)
		}

		fun newTradeMove(tradeOptions: List<ZPlayerName>): ZMove {
			return ZMove(type = ZMoveType.TRADE, list = tradeOptions)
		}

        fun newConsumeMove(equip: ZEquipment<*>, slot: ZEquipSlot?): ZMove {
            return ZMove(type = ZMoveType.CONSUME, equipment = equip, fromSlot = slot)
        }

        fun newEquipMove(equip: ZEquipment<*>, fromSlot: ZEquipSlot?, toSlot: ZEquipSlot, action: ZActionType? = null, player : ZPlayerName?=null): ZMove {
            return ZMove(type = ZMoveType.EQUIP, equipment = equip, fromSlot = fromSlot, toSlot = toSlot, action = action, character = player)
        }

        fun newKeepMove(equip: ZEquipment<*>): ZMove {
            return ZMove(type = ZMoveType.KEEP, equipment = equip)
        }

        fun newUnequippedMove(equip: ZEquipment<*>, slot: ZEquipSlot): ZMove {
	        return ZMove(type = ZMoveType.UNEQUIP, equipment = equip, fromSlot = slot)
        }

        fun newDisposeMove(equip: ZEquipment<*>, slot: ZEquipSlot?=null): ZMove {
            return ZMove(type = ZMoveType.DISPOSE, equipment = equip, fromSlot = slot?:equip.slot)
        }

        fun newGiveMove(taker: ZPlayerName, toGive: ZEquipment<*>, toSlot : ZEquipSlot?): ZMove {
            return ZMove(type = ZMoveType.GIVE, character = taker, equipment = toGive, fromSlot = toGive.slot, toSlot = toSlot)
        }

        fun newTakeMove(giver: ZPlayerName, toTake: ZEquipment<*>, toSlot : ZEquipSlot?): ZMove {
            return ZMove(type = ZMoveType.TAKE, character = giver, equipment = toTake, fromSlot = toTake.slot, toSlot = toSlot)
        }

	    fun newOrganizeTakeMove(taker: ZPlayerName, toTake: ZEquipment<*>, toSlot : ZEquipSlot?): ZMove {
		    return ZMove(type = ZMoveType.ORGANIZE_TAKE, character = taker, equipment = toTake, fromSlot = toTake.slot, toSlot = toSlot)
	    }

	    fun newObjectiveMove(zone: Int): ZMove {
            return ZMove(type = ZMoveType.TAKE_OBJECTIVE, integer = zone)
	    }

		fun newReloadMove(slot: ZWeapon): ZMove {
			return ZMove(type = ZMoveType.RELOAD, equipment = slot)
		}

		fun newPickupItemMove(takables: List<ZEquipment<*>>): ZMove {
			return ZMove(type = ZMoveType.PICKUP_ITEM, list = takables)
		}

		fun newPickupItemMove(character: ZPlayerName, equip: ZEquipment<*>, toSlot: ZEquipSlot): ZMove {
			return ZMove(character = character, type = ZMoveType.PICKUP_ITEM, equipment = equip, toSlot = toSlot)
		}

		fun newOrganizePickupMove(moves: List<ZMove>): ZMove {
			return ZMove(type = ZMoveType.ORGANIZE_PICKUP, list = moves)
		}

		fun newDropItemMove(items: List<ZEquipment<*>>): ZMove {
			return ZMove(type = ZMoveType.DROP_ITEM, list = items)
		}

		fun newDropItemMove(item: ZEquipment<*>): ZMove {
			return ZMove(type = ZMoveType.DROP_ITEM, equipment = item)
		}

		fun newWalkDirMove(dir: ZDir, action: ZActionType?): ZMove {
			return ZMove(type = ZMoveType.WALK_DIR, integer = dir.ordinal, action = action)
		}

		fun newSwitchActiveCharacter(): ZMove {
			return ZMove(type = ZMoveType.SWITCH_ACTIVE_CHARACTER)
		}

		fun newMakeNoiseMove(occupiedZone: Int): ZMove {
			return ZMove(type = ZMoveType.MAKE_NOISE, integer = occupiedZone)
		}

        fun newShoveMove(toZones: List<Int>): ZMove {
            return ZMove(type = ZMoveType.SHOVE, list = toZones)
        }

        fun newReRollMove(): ZMove {
            return ZMove(type = ZMoveType.REROLL)
        }

        fun newKeepRollMove(): ZMove {
            return ZMove(type = ZMoveType.KEEP_ROLL)
        }

        fun newEnchantMove(spells: List<ZSpell>): ZMove {
            return ZMove(type = ZMoveType.ENCHANT, list = spells)
        }

        fun newBornLeaderMove(options: List<ZPlayerName>): ZMove {
            return ZMove(type = ZMoveType.BORN_LEADER, list = options)
        }

        fun newBloodlustMeleeMove(zones: List<Int>, skill: ZSkill): ZMove {
            return ZMove(type = ZMoveType.BLOODLUST_MELEE, list = zones, skill = skill)
        }

        fun newBloodlustRangedMove(zones: List<Int>, skill: ZSkill): ZMove {
            return ZMove(type = ZMoveType.BLOODLUST_RANGED, list = zones, skill = skill)
        }

        fun newBloodlustMagicMove(zones: List<Int>, skill: ZSkill): ZMove {
            return ZMove(type = ZMoveType.BLOODLUST_MAGIC, list = zones, skill = skill)
        }

        fun newIgniteMove(ignitableZones: List<Int>): ZMove {
            return ZMove(type = ZMoveType.IGNITE, list = ignitableZones)
        }

	    fun newCloseSpawnPortal(zone: Int) : ZMove {
		    return ZMove(type = ZMoveType.CLOSE_SPAWN_PORTAL, integer = zone, skill = ZSkill.Hand_of_God)
	    }

	    fun newOrganize() : ZMove {
	    	return ZMove(type = ZMoveType.ORGANIZE)
	    }

	    fun newOrganizeDone() : ZMove {
		    return ZMove(type = ZMoveType.ORGANIZE_DONE)
	    }

		fun newOrganizeTrade(player: ZPlayerName): ZMove {
			return ZMove(type = ZMoveType.ORGANIZE_TRADE, character = player)
		}

		fun newOrganizeSlot(player: ZPlayerName, slot: ZEquipSlot?, moves: List<ZMove>): ZMove {
			return ZMove(
				type = ZMoveType.ORGANIZE_SLOT,
				fromSlot = slot,
				character = player,
				list = moves
			)
		}

		fun newMoveSiegeEngine(player: ZPlayerName, zoneOptions: List<Int>): ZMove {
			return ZMove(type = ZMoveType.SIEGE_ENGINE_MOVE, character = player, list = zoneOptions)
		}

		fun newFireCatapultScatterShot(player: ZPlayerName, zoneOptions: List<Int>): ZMove {
			return ZMove(
				type = ZMoveType.CATAPULT_FIRE_SCATTERSHOT,
				character = player,
				list = zoneOptions
			)
		}

		fun newFireCatapultGrapeShot(player: ZPlayerName, zoneOptions: List<Int>): ZMove {
			return ZMove(
				type = ZMoveType.CATAPULT_FIRE_GRAPESHOT,
				character = player,
				list = zoneOptions
			)
		}

		fun newFireCatapultBoulder(player: ZPlayerName, zoneOptions: List<Int>): ZMove {
			return ZMove(
				type = ZMoveType.CATAPULT_FIRE_BOULDER,
				character = player,
				list = zoneOptions
			)
		}


		fun newFireBallistaBolt(player: ZPlayerName, directionOptions: List<ZDir>): ZMove {
			return ZMove(
				type = ZMoveType.BALLISTA_FIRE_BOLT,
				character = player,
				list = directionOptions
			)
		}

		fun newFamiliarMove(player: ZPlayerName, familiar: ZFamiliar): ZMove {
			return ZMove(ZMoveType.FAMILIAR_MOVE, character = player, familiar = familiar.type)
		}

		init {
			addAllFields(ZMove::class.java)
		}
	}

	constructor(copy: ZMove, singleListElement: Any, text: String) : this(copy.type, copy.integer, copy.character, copy.equipment, copy.fromSlot, copy.toSlot, listOf(singleListElement), copy.skill, copy.action, text)
	constructor(copy: ZMove, singleListElement: Any, integer: Int, text: String? = null) : this(copy.type, integer, copy.character, copy.equipment, copy.fromSlot, copy.toSlot, listOf(singleListElement), copy.skill, copy.action, text)
	constructor(copy: ZMove, singleListElement: Any, character: ZPlayerName?, text: String) : this(copy.type, copy.integer, character, copy.equipment, copy.fromSlot, copy.toSlot, listOf(singleListElement), copy.skill, copy.action, text)
	constructor(copy: ZMove, singleListElement: Any, equipment: ZEquipment<*>?, text: String) : this(copy.type, copy.integer, copy.character, equipment, copy.fromSlot, copy.toSlot, listOf(singleListElement), copy.skill, copy.action, text)

	override fun equals(o: Any?): Boolean {
		if (this === o) return true
		if (o == null) return false
		if (o is ZMoveType) return type === o
		if (o !is ZMove) return false
		val zMove = o
		return (type === zMove.type && cc.lib.utils.isEqual(equipment, zMove.equipment)
			&& cc.lib.utils.isEqual(integer, zMove.integer)
			&& character === zMove.character
			&& fromSlot === zMove.fromSlot
			&& toSlot === zMove.toSlot
			&& skill === zMove.skill)
	}

	override fun hashCode(): Int {
		return type.hashCode()
	}

	override fun getTooltipText(): String? = equipment?.getTooltipText()
		?: skill?.getTooltipText()
		?: type.toolTipText
		?: character?.getTooltipText()

	override fun getLabel(): String {
		if (text != null)
			return text
		var label = type.getLabel()
		equipment?.let {
			label += " ${it.getLabel()}"
		}
		if (fromSlot != null) {
			label += " from ${fromSlot.getLabel()}"
		}
		if (toSlot != null)
			label += " to ${toSlot.getLabel()}"
		if (type == ZMoveType.ORGANIZE_TRADE) {
			character?.name?.let {
				label += " with $it"
			}
		}
		if (type == ZMoveType.PICKUP_ITEM && list?.size == 1) {
			label = "Pickup ${list[0]}"
		}
		if (type == ZMoveType.ENCHANT && list?.size == 1) {
			label = list[0].toString()
		}
		return label
	}

	override fun toString(): String {
		return super.toString()
	}
}